### PR TITLE
adds patient commands

### LINKED ISSUE #631: Model commands for patient aggregate

### PR DESCRIPTION
**Pull request by PaulKalho**
## Which issues does this pull request close?
closes #631

### CODE REVIEW

File: services/tasks-svc/internal/patient/aggregate/aggregate.go [hunk 1]
```diff
@@ -30,4 +31,81 @@ func LoadPatientAggregate(ctx context.Context, as hwes.AggregateStore, id uuid.U
...
+// Event handlers
+func (a *PatientAggregate) onPatientCreated(evt hwes.Event) error {
+	var payload patientEventsV1.PatientCreatedEvent
+	if err := evt.GetJsonData(&payload); err != nil {
+		return err
+	}
+
+	a.Patient.Notes = payload.Notes
+	a.Patient.HumanReadableIdentifier = payload.HumanReadableIdentifier
+	a.Patient.CreatedAt = evt.Timestamp
```
Comment (FoseFx) [inline services/tasks-svc/internal/patient/aggregate/aggregate.go:53]: also set updatedAt

File: services/tasks-svc/internal/patient/aggregate/aggregate.go [hunk 2]
```diff
@@ -30,4 +31,81 @@ func LoadPatientAggregate(ctx context.Context, as hwes.AggregateStore, id uuid.U
...
+	return nil
+}
+
+func (a *PatientAggregate) onBedUnassigned(_ hwes.Event) error {
+	a.Patient.BedID = uuid.NullUUID{UUID: uuid.Nil, Valid: false}
+	return nil
+}
+
+func (a *PatientAggregate) onPatientDischarged(_ hwes.Event) error {
+	a.Patient.IsDischarged = true
```
Comment (FoseFx) [inline services/tasks-svc/internal/patient/aggregate/aggregate.go:80]: set updatedAt

File: services/tasks-svc/internal/patient/aggregate/aggregate.go [hunk 3]
```diff
@@ -30,4 +31,81 @@ func LoadPatientAggregate(ctx context.Context, as hwes.AggregateStore, id uuid.U
...
+
+	a.Patient.HumanReadableIdentifier = payload.HumanReadableIdentifier
+	a.Patient.UpdatedAt = evt.Timestamp
+
+	return nil
+}
+
+func (a *PatientAggregate) onPatientReadmitted(_ hwes.Event) error {
+	a.Patient.IsDischarged = false
+	return nil
```
Comment (FoseFx) [inline services/tasks-svc/internal/patient/aggregate/aggregate.go:110]: set updatedAt

File: services/tasks-svc/internal/patient/aggregate/aggregate_test.go [hunk 4]
```diff
@@ -15,3 +18,137 @@ func MustApplyEvent(t *testing.T, aggregate hwes.Aggregate, newEvent func() (hwe
 		t.Error(err)
 	}
 }
+
+func TestPatientAggregate_CreatePatient(t *testing.T) {
```
Comment (FoseFx) [inline services/tasks-svc/internal/patient/aggregate/aggregate_test.go:22]: @MaxSchaefer is probably very happy about these, I, for one, think these kinds of tests are a burden to maintain and don't increase code confidence very much

File: services/tasks-svc/internal/patient/aggregate/actions.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+	// Therefore, we are using the same ID
+	id := a.GetID()
+	event, err := patientEventsV1.NewPatientCreatedEvent(a, id, humanReadableIdentifier, notes)
+	if err != nil {
+		return err
+	}
+	return a.Apply(event)
+}
+
+func (a *PatientAggregate) AssignBed(ctx context.Context, currentBedID uuid.NullUUID, newBedID uuid.UUID) error {
```
Comment (FoseFx) [inline services/tasks-svc/internal/patient/aggregate/actions.go:20]: why is `currentBedID` passed as an argument, is it not a field of `a.Patient`?

File: services/tasks-svc/internal/patient/aggregate/actions.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+func (a *PatientAggregate) UpdateHumanReadableIdentifier(ctx context.Context, currentHumanReadableIdentifier string, newHumanReadableIdentifier string) error {
+	event, err := patientEventsV1.NewHumanReadableIdentifierUpdatedEvent(a, currentHumanReadableIdentifier, newHumanReadableIdentifier)
+	if err != nil {
+		return err
+	}
+	return a.Apply(event)
+}
+
+func (a *PatientAggregate) UpdateNotes(ctx context.Context, currentNotes string, newNotes string) error {
```
Comment (FoseFx) [inline services/tasks-svc/internal/patient/aggregate/actions.go:44]: why is `currentNotes` passed as an argument, is it not a field of `a.Patient`?

File: services/tasks-svc/internal/patient/aggregate/actions.go [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+func (a *PatientAggregate) UnassignBed(ctx context.Context) error {
+	return a.Apply(patientEventsV1.NewBedUnassignedEvent(a))
+}
+
+func (a *PatientAggregate) DischargePatient(ctx context.Context) error {
+	return a.Apply(patientEventsV1.NewPatientDischargedEvent(a))
+}
+
+func (a *PatientAggregate) UpdateHumanReadableIdentifier(ctx context.Context, currentHumanReadableIdentifier string, newHumanReadableIdentifier string) error {
```
Comment (FoseFx) [inline services/tasks-svc/internal/patient/aggregate/actions.go:36]: why is `currentHumanReadableIdentifier` passed as an argument, is it not a field of `a.Patient`?

**Review by FoseFx — COMMENTED** [review]
Not done with this yet, but I already left some notes. In general, I am confused why events carry information about the previous state (e.g., previousBedID)

File: services/tasks-svc/internal/patient/aggregate/actions.go (hunk 5 above)
Comment (PaulKalho) [reply services/tasks-svc/internal/patient/aggregate/actions.go:20]: you're right, we can shift that down to the actions file. We also have to do that for the task aggregate.

**Comment by PaulKalho** [comment]
> Not done with this yet, but I already left some notes. In general, I am confused why events carry information about the previous state (e.g., previousBedID)

I've talked about that with @MaxSchaefer, we don't need the information obviously. But we decided to keep them in.

Also I'm not quite sure with my changes in https://github.com/helpwave/services/commit/8516ee8c624c815d83013783e3d5c8fe5c9c0340 .
I believe it's not very beautiful to pass `a` as well as `a.Task.` to the functions.
We can't use the TaskAggregate in Events due to circular Imports.

Under this circumstances, should we leave out previous State information from the events? Or can we deal with it differently? Wdyt @MaxSchaefer @FoseFx

File: services/tasks-svc/internal/patient/aggregate/aggregate_test.go (hunk 4 above)
Comment (MaxSchaefer) [reply services/tasks-svc/internal/patient/aggregate/aggregate_test.go:22]: Are you only referring to the aggregate tests here? Open to discuss :) (without blocking this pr)

**Comment by MaxSchaefer** [comment]
> I've talked about that with @MaxSchaefer, we don't need the information obviously. But we decided to keep them in.

We talked about this, yes. But what was the reasoning behind the decision the keep them in? I guess my initial idea was to revert a specific event but we can achieve this by utilizing our event sourcing :D

We can remove the previous data of the events.

**Review by FoseFx — APPROVED** [review]
:+1:

**Pull request merged**